Account Transfer Skips Account Chooser

fadidurah opened the pull request:
In this PR, i'm implementing the skipping of account chooser during account transfer requests.

changes:
- Pass account transfer token explicitly in InteractiveTokenCommandParameters
- Pass the transfer token as a separate field as part of the broker request
- Add a boolean to the broker request bundle to denote account transfer request
- Add boolean to BrokerInteractiveTokenCommandParameters to denote account transfer request

fadidurah committed:
add parameter to show if request is ATv2

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-parameter

fadidurah committed:
changelog

fadidurah committed:
draft changes for new api for ATv2

fadidurah committed:
add back methodname

fadidurah committed:
adjust def

fadidurah committed:
revert some changes

fadidurah committed:
explicit parameter creation

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
fix pmd

fadidurah committed:
add check for slk

fadidurah committed:
return method tag

fadidurah committed:
return getBrokerAuthorizationIntent

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
comment

fadidurah committed:
add error for missing slk

fadidurah committed:
remove test code

fadidurah committed:
Update common4j/src/main/com/microsoft/identity/common/java/exception/ClientException.java

Co-authored-by: Shahzaib <37125644+shahzaibj@users.noreply.github.com>

fadidurah committed:
merge other branch

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
adjust URI

fadidurah committed:
changelog

fadidurah commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1599,6 +1600,12 @@ public String getMsalVersion(){
 
+        /**
+         * URI Path constant for MSAL-to-Broker acquireTokenInteractive (for ATv2) request using ContentProvider.
+         *  TODO: Not really sure how this works, this seems right
```
Ping

fadidurah replied:
No longer applicable

fadidurah commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -415,6 +421,109 @@ public void onReceive(@NonNull PropertyBag propertyBag) {
+
+            final String negotiatedBrokerProtocolVersion = interactiveATv2RequestIntent.getStringExtra(NEGOTIATED_BP_VERSION_KEY);
+
+            // TODO: Is this accurate? BrokerAccountDataManager contains MSA accounts
```
Ping

mohitc1 replied:
This should not be needed anymore. This was before PRTv3 when MSA accounts where not supported in broker.

fadidurah commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -484,6 +593,75 @@ public void putValueInSuccessEvent(final @NonNull ApiEndEvent event, final @NonN
+
+                    // TODO: Confused, why is this called "extractResultBundle", but we're actually
+                    //  returning an intent where we create the bundle for the request?
+                    //  Is this the result of performPrerequisites with broker?
```
Ping

mohitc1 replied:
This for extract the result obtained from broker (resultBundle) into object you want. In this case it's Intent object (BrokerOperation<Intent>()  {} above)

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1536,6 +1536,7 @@ public static final class BrokerContentProvider {
         public enum API {
             MSAL_HELLO(MSAL_HELLO_PATH, null, VERSION_3),
             ACQUIRE_TOKEN_INTERACTIVE(MSAL_ACQUIRE_TOKEN_INTERACTIVE_PATH, null, VERSION_3),
+            ACQUIRE_TOKEN_ATV2_INTERACTIVE(MSAL_ACQUIRE_TOKEN_ATV2_INTERACTIVE_PATH, null, VERSION_3),
```
>VERSION_3 [](http://example.com/codeflow?start=91&length=9)

Do we need to update this?

fadidurah replied:
I'm honestly not sure, was hoping someone more familiar with this part of the code would give a definitive answer 😅

shahzaibj replied:
I'm pretty sure this needs to be updated because older brokers are not going to understand ATv2 API.

mohitc1 replied:
yes, just that this parameter is not in used today but can be put to use when making call from BrokerMsalController

fadidurah replied:
Is this not already covered by bumping the broker protocol version to 16?

mohitc1 replied:
You may update here as well (don't see any effect of it though)

mohitc1 commented on common/src/main/aidl/com/microsoft/identity/client/IMicrosoftAuthService.aidl:
```diff
@@ -38,6 +38,8 @@ interface IMicrosoftAuthService {
 
     Intent getIntentForInteractiveRequest();
 
+    Intent getIntentForATv2InteractiveRequest();
```
>ATv2 [](http://example.com/codeflow?start=23&length=4)

AccountTransferV2 instead of ATv2?

fadidurah replied:
Sure thing

fadidurah replied:
renamed all references to ATv2 in both PRs

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/commands/parameters/ATv2TokenCommandParameters.java:
```diff
@@ -0,0 +1,39 @@
+@Getter
+@EqualsAndHashCode(callSuper = true)
+@SuperBuilder(toBuilder = true)
+public class ATv2TokenCommandParameters extends InteractiveTokenCommandParameters {
```
Why not we define slk as a required parameter here? You'd get parameter validation while request creation itself and forces a technical constraint on Atv2 parameters

fadidurah replied:
SLK is already being passed as an extra query parameter in the base token command parameters. I'd rather not extract it into its own field because that would mean we'd need to add another code path in acquirePrt to manually pass in slk in ATv2 scenarios to the authorize endpoint. This functionality is already present if SLK is in the extra query parameters field, I don't think we should ask oneauth to add it in 2 places within the same parameters object.

We could do it in a way where they only pass it in as a separate field in ATv2TokenCommandParameters, and we do some check in ATv2Activity to add it into the extra query parameter if it's not already there before sending the request through PrtController, but i feel this is adding more complexity that we probably don't need and is more or less still doing the parameter validation anyway.

fadidurah replied:
After some design changes, we've updated this to include slk as a direct parameter in this class.

fadidurah committed:
AccountTransferV2 instead of ATv2

fadidurah committed:
accountTransferV2 instead of atv2

fadidurah committed:
merge dev

fadidurah committed:
updated design, pick account in activity rather than controller

fadidurah committed:
t add .
git commit -m ":wq
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
unused method

fadidurah committed:
bump protocol version

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -415,6 +417,109 @@ public void onReceive(@NonNull PropertyBag propertyBag) {
+            // If the request is from MSALCPP , OAuth2TokenCache will be null.
+            if (parameters.getOAuth2TokenCache() != null && !BrokerProtocolVersionUtil.canSupportMsaAccountsInBroker(negotiatedBrokerProtocolVersion)) {
+                saveMsaAccountToCache(resultBundle, (MsalOAuth2TokenCache) parameters.getOAuth2TokenCache());
+            }
```
This should not be needed anymore. This was before PRTv3 when MSA accounts where not supported in broker.

fadidurah replied:
What does "this" refer to, the whole if statement?

mohitc1 replied:
saveMsaAccountToCache part

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -260,6 +281,23 @@ public Bundle getRequestBundleForAcquireTokenInteractive(@NonNull final Interact
+        );
+    }
+
     /**
```
Add UTs for these

fadidurah committed:
update dev

fadidurah committed:
new test

fadidurah committed:
add a Unit test

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -90,12 +91,33 @@ public BrokerRequest brokerRequestFromAcquireTokenParameters(@NonNull final Inte
                 QueryParamsAdapter._toJson(parameters.getExtraOptions()) : null;
-        final BrokerRequest brokerRequest = BrokerRequest.builder()
+
+        final BrokerRequest.BrokerRequestBuilder brokerRequest = BrokerRequest.builder()
```
>brokerRequest [](http://example.com/codeflow?start=49&length=13)

nit: rename to brokerRequestBuilder

mohitc1 reviewed (approved):
Approved with suggestions

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API
t status

shahzaibj commented on common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java:
```diff
@@ -123,6 +124,10 @@ public static String getDelimitedDefaultScopeString() {
     public abstract AcquireTokenResult acquireToken(final InteractiveTokenCommandParameters request)
             throws Exception;
 
+    public AcquireTokenResult acquireTokenForAccountTransferV2(final AccountTransferV2TokenCommandParameters request) throws Exception {
```
Why not declare it abstract here like we are doing with the rest of the methods?

fadidurah replied:
If i declare it abstract here, i'd need to define it in every class that extends BaseController, and have them all (except BrokerMsalController) throw an unsupported method exception. This seemed cleaner to do it this way, and just override it in BrokerMsalController.

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
rename to remove reference to v2

fadidurah committed:
fix spotbug

fadidurah committed:
remove uninstended change

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API

fadidurah committed:
merge dev

rpdome commented on common/src/main/aidl/com/microsoft/identity/client/IMicrosoftAuthService.aidl:
```diff
@@ -38,6 +38,8 @@ interface IMicrosoftAuthService {
 
     Intent getIntentForInteractiveRequest();
 
+    Intent getIntentForAccountTransferInteractiveRequest();
```
add it at the bottom. 

there might be some weird side effects if the order of aidl on broker and MSAL/OneAuth side mismatches.

fadidurah replied:
Done

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java:
```diff
@@ -98,6 +98,20 @@ public MicrosoftAuthClient(@NonNull final Context context,
+                final Intent accountTransferIntent = microsoftAuthService.getIntentForAccountTransferInteractiveRequest();
+                final Bundle accountTransferBundle = accountTransferIntent.getExtras();
+
+                //older brokers (pre-ContentProvider) are ONLY sending these values in the intent itself.
```
this msg could be confusing - since this operation is just added (it's not pre-ContentProvider)

you actually can just expose 
GetBundleForAccountTransferInteractiveRequestMsalBrokerOperation
on the Broker side, and return the resulted bundle as is.

The code will be cleaner on the Broker side, and you wouldn't need to do extra conversion here.

fadidurah replied:
No longer applicable

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -484,6 +575,70 @@ public void putValueInSuccessEvent(final @NonNull ApiEndEvent event, final @NonN
+                    @Override
+                    public void performPrerequisites(final @NonNull IIpcStrategy strategy) throws BaseException {
+                        verifyTokenParametersAreSupported(parameters);
+                        negotiatedBrokerProtocolVersion = hello(strategy, parameters.getRequiredBrokerProtocolVersion());
```
if the negotiatedBrokerProtocolVersion is lower than the new version you just added, it means the broker does NOT support this operation.

In that case, you might want to fail the request early. (no need to make the request to the broker because it'll just throw "not supported" back)

fadidurah replied:
So, in the parameters for ATv2, OneAuth will send in 16.0 protocol version. wouldn't that already take care of this issue? We'd get an error result from the hello operation in that case, right?

rpdome replied:
Short answer: you're right.

Long answer: 
1. OneAuth would send a hello request with 16.0 as the *minimum* protocol version
2. If the broker doesn't support 16.0, the following block will be hit
```Java
        // Broker is older than the minimum version that the client requires.
        Logger.warn(
                methodTag,
                "Connection DISALLOWED"
                        + "\n"
                        + "Broker only supports protocol version up to: " + brokerMaxProtocolVersion
                        + " but client requires: " + clientMinProtocolVersion
        );

        final Bundle resultBundle = new Bundle();
        resultBundle.putString(HELLO_ERROR_CODE, ErrorStrings.UNSUPPORTED_BROKER_VERSION_ERROR_CODE);
        resultBundle.putString(HELLO_ERROR_MESSAGE, ErrorStrings.UNSUPPORTED_BROKER_VERSION_ERROR_MESSAGE);
        return resultBundle;
```
3. BrokerMsalController.hello() will invoke verifyHelloFromResultBundle(), and that method will convert an error into an exception.

fadidurah replied:
No longer applicable

fadidurah committed:
merge dev

fadidurah committed:
reorder method

fadidurah committed:
remove comment

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapterTests.java:
```diff
@@ -75,4 +85,42 @@ public void test_getRequestBundleForSsoToken() {
+                .prompt(OpenIdConnectPromptParameter.SELECT_ACCOUNT)
+                .build();
+
+        MsalBrokerRequestAdapter msalBrokerRequestAdapter = new MsalBrokerRequestAdapter();
```
nit: final

fadidurah replied:
No longer applicable

rpdome commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1536,6 +1537,7 @@ public static final class BrokerContentProvider {
         public enum API {
             MSAL_HELLO(MSAL_HELLO_PATH, null, VERSION_3),
             ACQUIRE_TOKEN_INTERACTIVE(MSAL_ACQUIRE_TOKEN_INTERACTIVE_PATH, null, VERSION_3),
+            ACQUIRE_TOKEN_ACCOUNT_TRANSFER_INTERACTIVE(MSAL_ACQUIRE_TOKEN_ACCOUNT_TRANSFER_INTERACTIVE_PATH, null, VERSION_8),
```
protocol version 14.0?

https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/%5BAndroid%5D%20Broker%20API/broker_protocol_versions.md&_a=preview

(not sure if this docs is updated btw)

fadidurah replied:
No longer needed

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -415,6 +414,98 @@ public void onReceive(@NonNull PropertyBag propertyBag) {
+     */
+    @Override
+    public AcquireTokenResult acquireTokenForAccountTransfer(final @NonNull AccountTransferTokenCommandParameters parameters) throws BaseException, ExecutionException, InterruptedException {
+        final String methodTag = TAG + ":acquireTokenForAccountTransfer";
```
is there any way to avoid duplicating the code in this method?

fadidurah replied:
No longer needed

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -484,6 +575,70 @@ public void putValueInSuccessEvent(final @NonNull ApiEndEvent event, final @NonN
+                        return new BrokerOperationBundle(
+                                MSAL_GET_INTENT_FOR_ACCOUNT_TRANSFER_INTERACTIVE_REQUEST,
+                                mActiveBrokerPackageName,
+                                null);
```
alternatively, you can create a bundle with a flag here.

on the broker side, if the flag exists, return the AccountTransfer activity class. 
otherwise, return AccountChooser.

If you do it this way, then the change on msal side would be minimal. you'll get a lot of things for free.

(you can pretty much reuse the existing acquire token interactive code in this class. The only difference would be in this getBundle() method)

rpdome replied:
could you please remind me again what's the reason of opening a new ipc interface/new activity?

fadidurah replied:
No longer needed

rpdome commented:
Feel free to ping me / set up a meeting to expedite the review process.

fadidurah committed:
remove unneeded changes

fadidurah committed:
more unneeded changes

fadidurah committed:
merge dev

fadidurah committed:
revert adapter changes

fadidurah committed:
more unneeded changes removed

fadidurah committed:
pass boolean for account transfer

fadidurah committed:
remove more changes

fadidurah committed:
add boolean to bundle

fadidurah committed:
changelog

fadidurah committed:
nits

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -310,10 +310,13 @@ private String tryGetNegotiatedProtocolVersionFromHelloCache(
             throws BaseException, InterruptedException, ExecutionException {
         final String methodTag = TAG + ":acquireToken";
 
+        // This is for testing only, will remove once validation is complete. Same parameter renaming
```
remove this.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerMsalController.java:
```diff
@@ -387,10 +390,7 @@ public void onReceive(@NonNull PropertyBag propertyBag) {
             final Bundle resultBundle = mBrokerResultFuture.get();
 
             final String negotiatedBrokerProtocolVersion = interactiveRequestIntent.getStringExtra(NEGOTIATED_BP_VERSION_KEY);
-            // For MSA Accounts Broker doesn't save the accounts, instead it just passes the result along,
```
nit: why removing this?

fadidurah replied:
I think this is an old comment, because Broker does save the MSA accounts. This detail is actually important to my implementation because we query broker storage for msa accounts.

fadidurah replied:
@mohitc1 Let me know if i'm understanding this wrong. I can add bnack the comment if it's still relevant.

fadidurah replied:
I've returned the comment, we can remove it later if necessary.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/broker/BrokerRequest.java:
```diff
@@ -253,4 +253,8 @@ private static final class SerializedNames {
+
+    @Nullable
+    @SerializedName(SerializedNames.ACCOUNT_TRANSFER_TOKEN)
+    private String mAccountTransferToken;
```
sync with @mohitc1.

If he's planning to deserialize the BrokerRequest in AccountChooser.shouldDisplayAccountList(), then you might not need this. you can use BrokerInteractiveTokenCommandParameters.isAccountTransferRequest() instead.

fadidurah replied:
My change was to allow the check without deserializing, Is it possible to do something similar in your implementation @mohitc1 ? Or do we feel we should pivot to deserializing the brokerRequest inside the account chooser.

fadidurah replied:
I pulled mohit's changes and used the earlier serializing infra

fadidurah committed:
pull dev

fadidurah committed:
changelog

fadidurah committed:
don't need bundle boolean

fadidurah committed:
revert testing change

fadidurah committed:
merge dev

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-common-for-android into fadi/ATv2-API
